src/pyhf/cli/complete.py

```diff
@@ -1,30 +1,65 @@
+"""Shell completions for pyhf."""
 
 import click
 
-try:
-    import click_completion
 
-    click_completion.init()
+@click.command(help="Generate shell completion code.", name="completions")
+@click.argument(
+    "shell",
+    required=False,
+    type=click.Choice(["bash", "zsh", "fish"], case_sensitive=False),
+)
+def cli(shell):
+    """Generate shell completion code for various shells.
 
-    @click.command(help='Generate shell completion code.', name='completions')
-    @click.argument(
-        'shell',
-        required=False,
-        type=click_completion.DocumentedChoice(click_completion.core.shells),
-    )
-    def cli(shell):
-        '''Generate shell completion code for various shells.'''
-        click.echo(click_completion.core.get_code(shell, prog_name='pyhf'))
-
-except ImportError:
-
-    @click.command(help='Generate shell completion code.', name='completions')
-    @click.argument('shell', default=None)
-    def cli(shell):
-        """Placeholder for shell completion code generatioon function if necessary dependency is missing."""
-        click.secho(
-            "This requires the click_completion module.\n"
-            "You can install it with the shellcomplete extra:\n"
-            "python -m pip install 'pyhf[shellcomplete]'"
+    Supported shells: bash, zsh, fish
+
+    To enable completion, run the appropriate command for your shell:
+
+    \b
+    Bash:
+      mkdir -p ~/.completions
+      _PYHF_COMPLETE=bash_source pyhf > ~/.completions/pyhf-complete.sh
+      echo -e "\n. ~/.completions/pyhf-complete.sh" >> ~/.bashrc
+
+    \b
+    Zsh:
+      mkdir -p ~/.completions
+      _PYHF_COMPLETE=zsh_source pyhf > ~/.completions/pyhf-complete.zsh
+      echo -e "\n. ~/.completions/pyhf-complete.zsh" >> ~/.zshrc
+
+    \b
+    Fish:
+      _PYHF_COMPLETE=fish_source pyhf >> ~/.config/fish/completions/pyhf.fish
+    """
+    if shell is None:
+        click.echo(cli.get_help(click.Context(cli)))
+        return
+
+    click.echo(f"To enable {shell} completion for pyhf run in your {shell} shell:\n")
+
+    instructions = {
+        "bash": (
+            "mkdir -p ~/.completions\n"
+            "_PYHF_COMPLETE=bash_source pyhf > ~/.completions/pyhf-complete.sh\n"
+            'echo -e "\\n. ~/.completions/pyhf-complete.sh" >> ~/.bashrc\n'
+        ),
+        "zsh": (
+            "mkdir -p ~/.completions\n"
+            "_PYHF_COMPLETE=zsh_source pyhf > ~/.completions/pyhf-complete.zsh\n"
+            'echo -e "\\n. ~/.completions/pyhf-complete.zsh" >> ~/.zshrc\n'
+        ),
+        "fish": (
+            "_PYHF_COMPLETE=fish_source pyhf >> ~/.config/fish/completions/pyhf.fish\n"
+        ),
+    }
+    click.echo(
+        click.style(
+            instructions[shell],
+            bold=True,
         )
+    )
+    click.echo(
+        "and then source your shell configuration or restart your shell."
+        + "\nIt is recommended to press tab twice (<TAB><TAB>) to engage the completion during use."
```

Comment:
I've learned that while still a bit slow, some of my complaints were human user error.

e.g. I should be been doing double tab (`<TAB><TAB>`) all along.

```console
$ pyhf <TAB><TAB>
cls          combine      completions  contrib      digest       fit          inspect      json2xml     patchset     prune        rename       sort         xml2json
$ pyhf fit -<TAB><TAB>
--output-file  --measurement  -p             --patch        --value        --backend      --optimizer    --optconf      --help         -h 
```

Reply:
do we care if double-tabbing is slow? it can be slow for some applications, but as long as the normal single-tab to complete an argument like `--outpu<TAB>` is fast, then that's ok?

Reply:
@kratsg Probably not, but `pyhf fi<TAB>` takes about a second to complete to `pyhf fit` for me. So none of these are super fast.